### PR TITLE
[SPARK-39532][INFRA] Move checkout and sync steps into re-usable action

### PR DESCRIPTION
**Pull request by EnricoMi**
### What changes were proposed in this pull request?
Steps "Checkout Spark repository" and "Sync the current branch with the latest in Apache Spark" are defined 9 times in `build-and-test.yml`. This moves these two steps into a composite action that is than reused.

### Why are the changes needed?
Simplifies workflow file and reduces code duplication.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Tested in fork CI.

### CODE REVIEW

**Comment by EnricoMi** [comment]
@HyukjinKwon here is some first simplification of the `build-and-test` workflow.

**Review by HyukjinKwon — APPROVED** [review]
Looks nice. LGTM if it works

File: .github/actions/sync-fork-with-upstream/action.yml
Comment (HyukjinKwon) [inline .github/actions/sync-fork-with-upstream/action.yml:18]: quick question: what is this?

**Comment by EnricoMi** [comment]
This is not applicable to Apache Spark CI as it first checksout Spark repo, then the fork repo, while this action (until merged into upstream repo) only becomes available after the sync step, which is main part of the composite action.

In other words, this action cannot be executed until it has been merged into master. This makes it also difficult to contribute to.

**Comment by HyukjinKwon** [comment]
uhoh, right. thanks for explanation though :-).

File: .github/actions/sync-fork-with-upstream/action.yml
Comment (EnricoMi) [reply .github/actions/sync-fork-with-upstream/action.yml:18]: This is defines the icon of the action for UI and marketplace purposes, for instance the blue circle with the triangle at https://github.com/marketplace/actions/checkout

Not sure if this is required, does not make much sense if action is not published.

**Pull request closed without merging**
